runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java

```diff
@@ -92,4 +138,36 @@ public void readSoruce(ProcessContext ctxt) throws IOException {
       }
     }
   }
+
+  /**
+   * A {@link Coder} for {@link BoundedSource}s that wraps a {@link SerializableCoder}. We cannot
+   * safely use an unwrapped SerializableCoder because
+   * {@link SerializableCoder#structuralValue(Serializable)} assumes that coded elements support
+   * object equality. By default, Coders compare equality by serialized bytes, which we want in
+   * this case. It is usually safe to depend on coded representation here because we only compare
```

Comment:
"usually safe..." isn't really necessary.

Probably worth noting that we usually expect Source `equals` methods to not be defined.

Reply:
Fixed.